dtm0/fop.c

```diff
@@ -244,154 +232,177 @@ static size_t dtm0_fom_locality(const struct m0_fom *fom)
-	if (fop == NULL)
-		return M0_ERR(-ENOMEM);
-
-	item              = &fop->f_item;
-	item->ri_ops      = &dtm0_req_fop_rpc_item_ops;
-	item->ri_session  = session;
-	item->ri_prio     = M0_RPC_ITEM_PRIO_MID;
-	item->ri_deadline = M0_TIME_IMMEDIATELY;
-
-	req               = m0_fop_data(fop);
-	req->dtr_msg      = msg_type;
-	rc = m0_dtm0_tx_desc_copy(txd, &req->dtr_txr) ?: m0_rpc_post(item);
-	if (rc == 0) {
-		phase_sm_id = m0_sm_id_get(&fom->fo_sm_phase);
-		rpc_sm_id   = m0_sm_id_get(&item->ri_sm);
-		M0_ADDB2_ADD(M0_AVI_FOM_TO_TX, phase_sm_id, rpc_sm_id);
-
-		m0_fop_put_lock(fop);
-
-		M0_LOG(M0_DEBUG, "Sent %d msg " FID_F " -> " FID_F, msg_type,
-		       FID_P(&dtms->dos_generic.rs_service_fid), FID_P(tgt));
-	}
-	return M0_RC(rc);
+	return m0_dtm0_req_post(dtms,
+				&((struct dtm0_req_fop) {
+				  .dtr_msg = msg_type,
+				  .dtr_txr = *txd
+				  }),
+				tgt, fom, false);
 }
 
 M0_INTERNAL int m0_dtm0_logrec_update(struct m0_be_dtm0_log  *log,
 				      struct m0_be_tx        *tx,
 				      struct m0_dtm0_tx_desc *txd,
-				      struct m0_buf          *pyld)
+				      struct m0_buf          *payload)
 {
 	int rc;
 
 	M0_ENTRY();
 
 	m0_mutex_lock(&log->dl_lock);
-	rc = m0_be_dtm0_log_update(log, tx, txd, pyld);
+	rc = m0_be_dtm0_log_update(log, tx, txd, payload);
 	m0_mutex_unlock(&log->dl_lock);
 
 	return M0_RC(rc);
 }
 
-M0_INTERNAL int m0_dtm0_on_committed(struct m0_fom                *fom,
-				     const struct m0_dtm0_tx_desc *txd)
+M0_INTERNAL int m0_dtm0_on_committed(struct m0_fom            *fom,
+				     const struct m0_dtm0_tid *id)
 {
 	struct m0_dtm0_service *dtms = m0_dtm0_service_find(
-		fom->fo_service->rs_reqh);
-	struct m0_dtm0_tx_desc  msg = {};
+					fom->fo_service->rs_reqh);
+	struct m0_be_dtm0_log  *log = dtms->dos_log;
+	struct m0_dtm0_log_rec *rec;
+	struct m0_dtm0_tx_desc  txd;
+	struct m0_fid          *target;
+	struct m0_fid          *source = &dtms->dos_generic.rs_service_fid;
 	int                     rc;
 	int                     i;
 
-	rc = m0_dtm0_tx_desc_copy(txd, &msg);
+	M0_PRE(log != NULL);
+	M0_PRE(log->dl_is_persistent);
+
+	M0_ENTRY();
+
+	m0_mutex_lock(&log->dl_lock);
+	/* Get the latest state of the log record. */
+	rec = m0_be_dtm0_log_find(log, id);
+	M0_ASSERT_INFO(rec != NULL, "Log record must be inserted into the log "
+		       "in cas_fom_tick().");
+	rc = m0_dtm0_tx_desc_copy(&rec->dlr_txd, &txd);
+	m0_mutex_unlock(&log->dl_lock);
+
 	if (rc != 0)
-		return M0_ERR(rc);
-
-	/*
-	 * TODO: This change will be done by DTM0 log because
-	 * it should log the entry "pa == self" with PERSISTENT state
-	 * set.
-	 */
-	for (i = 0; i < msg.dtd_ps.dtp_nr; ++i) {
-		if (m0_fid_eq(&msg.dtd_ps.dtp_pa[i].p_fid,
-			      &dtms->dos_generic.rs_service_fid)) {
-			msg.dtd_ps.dtp_pa[i].p_state =
-				max_check(msg.dtd_ps.dtp_pa[i].p_state,
-					  (uint32_t) M0_DTPS_PERSISTENT);
+		goto out;
+
+	for (i = 0; i < txd.dtd_ps.dtp_nr; ++i) {
+		target = &txd.dtd_ps.dtp_pa[i].p_fid;
```

Comment:
I have moved this into the loop to have one single call: if we encounter our own FID then we are using the originator FID instead of our own. However, I'm not sure that it was a good idea.